### ISSUE TITLE
[1.18.2] Server Hang with Ice and Fire and Modernfix

### ISSUE THREAD

**Issue by Reshy** [CCr4ft3r/DimensionalThreading-Reforged#8]
"dimthread_overworld" prio=5 Id=187 WAITING on java.util.concurrent.CompletableFuture$Signaller@48c031b7
	at java.base@17.0.1/jdk.internal.misc.Unsafe.park(Native Method)
	-  waiting on java.util.concurrent.CompletableFuture$Signaller@48c031b7
	at java.base@17.0.1/java.util.concurrent.locks.LockSupport.park(LockSupport.java:211)
	at java.base@17.0.1/java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1864)
	at java.base@17.0.1/java.util.concurrent.ForkJoinPool.unmanagedBlock(ForkJoinPool.java:3463)
	at java.base@17.0.1/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3434)
	at java.base@17.0.1/java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1898)
	at java.base@17.0.1/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2117)
	at TRANSFORMER/iceandfire@2.1.12-1.18.2-beta2/com.github.alexthe666.iceandfire.pathfinding.raycoms.Pathfinding$IafThreadFactory.newThread(Pathfinding.java:96)
	...

	Number of locked synchronizers = 1
	- java.util.concurrent.ThreadPoolExecutor$Worker@1c3179b3

extended trace:
	at java.base@17.0.1/java.util.concurrent.ThreadPoolExecutor$Worker.<init>(ThreadPoolExecutor.java:630)
	at java.base@17.0.1/java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:920)
	at java.base@17.0.1/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1353)
	at java.base@17.0.1/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:145)
	at TRANSFORMER/iceandfire@2.1.12-1.18.2-beta2/com.github.alexthe666.iceandfire.pathfinding.raycoms.PathResult.startJob(PathResult.java:170)
	at TRANSFORMER/iceandfire@2.1.12-1.18.2-beta2/com.github.alexthe666.iceandfire.pathfinding.raycoms.AdvancedPathNavigate.setPathJob(AdvancedPathNavigate.java:244)
	at TRANSFORMER/iceandfire@2.1.12-1.18.2-beta2/com.github.alexthe666.iceandfire.pathfinding.raycoms.AdvancedPathNavigate.moveToXYZ(AdvancedPathNavigate.java:386)
	at TRANSFORMER/iceandfire@2.1.12-1.18.2-beta2/com.github.alexthe666.iceandfire.pathfinding.raycoms.AdvancedPathNavigate.m_26519_(AdvancedPathNavigate.java:477)
	at TRANSFORMER/iceandfire@2.1.12-1.18.2-beta2/com.github.alexthe666.iceandfire.entity.ai.HippogryphAIWander.m_8056_(HippogryphAIWander.java:63)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.entity.ai.goal.WrappedGoal.m_8056_(WrappedGoal.java:41)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.entity.ai.goal.GoalSelector.m_25373_(GoalSelector.java:106)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.entity.Mob.m_6140_(Mob.java:701)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.entity.LivingEntity.m_8107_(LivingEntity.java:2506)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.entity.Mob.m_8107_(Mob.java:498)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.entity.AgeableMob.m_8107_(AgeableMob.java:127)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.entity.animal.Animal.m_8107_(Animal.java:54)
	at TRANSFORMER/iceandfire@2.1.12-1.18.2-beta2/com.github.alexthe666.iceandfire.entity.EntityHippogryph.m_8107_(EntityHippogryph.java:808)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.entity.LivingEntity.m_8119_(LivingEntity.java:2253)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.entity.Mob.m_8119_(Mob.java:310)
	at TRANSFORMER/iceandfire@2.1.12-1.18.2-beta2/com.github.alexthe666.iceandfire.entity.EntityHippogryph.m_8119_(EntityHippogryph.java:978)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.level.ServerLevel.m_8647_(ServerLevel.java:652)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.level.ServerLevel$$Lambda$43097/0x000000080463a230.accept(Unknown Source)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.level.Level.m_46653_(Level.java:7074)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.level.ServerLevel.m_184063_(ServerLevel.java:319)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.level.ServerLevel$$Lambda$43096/0x000000080463a000.accept(Unknown Source)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.world.level.entity.EntityTickList.m_156910_(EntityTickList.java:54)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.level.ServerLevel.m_8793_(ServerLevel.java:299)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.MinecraftServer.mde124e5$lambda$tickWorlds$0$3(MinecraftServer.java:5976)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.MinecraftServer$$Lambda$43056/0x000000080462a120.run(Unknown Source)
	at TRANSFORMER/dimthread@1.18.2-1.2.8/wearblackallday.dimthread.DimThread.swapThreadsAndRun(DimThread.java:51)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.MinecraftServer.mde124e5$lambda$tickWorlds$1$2(MinecraftServer.java:5973)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.MinecraftServer$$Lambda$43051/0x0000000804629638.accept(Unknown Source)
	at TRANSFORMER/JavaUtils.bd58640372/wearblackallday.util.ThreadPool.lambda$execute$1(ThreadPool.java:49)
	at TRANSFORMER/JavaUtils.bd58640372/wearblackallday.util.ThreadPool$$Lambda$43053/0x0000000804629a98.run(Unknown Source)
	at TRANSFORMER/JavaUtils.bd58640372/wearblackallday.util.ThreadPool.lambda$execute$0(ThreadPool.java:43)
	at TRANSFORMER/JavaUtils.bd58640372/wearblackallday.util.ThreadPool$$Lambda$43054/0x0000000804629cb8.run(Unknown Source)
	at java.base@17.0.1/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base@17.0.1/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base@17.0.1/java.lang.Thread.run(Thread.java:833)

"dimthread_the_nether" prio=5 Id=188 BLOCKED on java.lang.Class@1cd83f8c owned by "dimthread_overworld" Id=187
	at TRANSFORMER/dimthread@1.18.2-1.2.8/wearblackallday.dimthread.DimThread.swapThreadsAndRun(DimThread.java:43)
	-  blocked on java.lang.Class@1cd83f8c
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.MinecraftServer.mde124e5$lambda$tickWorlds$1$2(MinecraftServer.java:5973)
	at TRANSFORMER/minecraft@1.18.2/net.minecraft.server.MinecraftServer$$Lambda$43051/0x0000000804629638.accept(Unknown Source)
	at TRANSFORMER/JavaUtils.bd58640372/wearblackallday.util.ThreadPool.lambda$execute$1(ThreadPool.java:49)
	at TRANSFORMER/JavaUtils.bd58640372/wearblackallday.util.ThreadPool$$Lambda$43053/0x0000000804629a98.run(Unknown Source)
	at TRANSFORMER/JavaUtils.bd58640372/wearblackallday.util.ThreadPool.lambda$execute$0(ThreadPool.java:43)
	at TRANSFORMER/JavaUtils.bd58640372/wearblackallday.util.ThreadPool$$Lambda$43054/0x0000000804629cb8.run(Unknown Source)
	at java.base@17.0.1/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	...

	Number of locked synchronizers = 1
	- java.util.concurrent.ThreadPoolExecutor$Worker@1ceb7d9d

Full thread dump:

https://pastebin.com/UWX87DuY

**Comment by CCr4ft3r** [comment]
Isn't this a duplicate of #7 ?

**Comment by Reshy** [comment]
No, this happened in the overworld.

**Comment by CCr4ft3r** [comment]
You probably need to remove Modernfix or disable the chunk deadlock mixin.

**Comment by embeddedt** [comment]
I think the most likely cause is that [this part of Ice and Fire's code](https://github.com/AlexModGuy/Ice_and_Fire/blob/5606aa43404c7c80061050ae6a9f5dccf041ba7a/src/main/java/com/github/alexthe666/iceandfire/pathfinding/raycoms/Pathfinding.java#L93-L97) is probably assuming a single server thread, which is now not true. In general I'd expect content mods to have issues with Dimensional Threading without explicit patches as this assumption is fairly common.

**Comment by CCr4ft3r** [comment]
@embeddedt Thx for your time to find the cause. 
@Reshy Unfortunately thats the case for a lot of other mods. So the only recommended way in my opinion would be to request DimThread support on Ice And Fire. It would be theoretically possible to fix it via mixins on DimThread side but could possibly break after switching to a newer version.
